### PR TITLE
Redirect properly after sign-in

### LINKED ISSUE #3882: Post-login should take you back to your previous page
As a user of Builder who has previously been authenticated, when I'm on a package page and login again, I should be taken back to that page after logging in. At least, that's what I've been expecting 😉

### PR DESCRIPTION
**Pull request by cnunciato**
This change adds a few new behaviors related to sign-in:

* When your session expires and you're asked to sign in again, once you do, you'll be redirected to the page you were trying to access

* If you aren't signed in and you visit a page that requires authentication, you'll be redirected to that page after you sign in

* If errors occur during the sign-in process (e.g., GitHub responds in disappointing ways), you'll be directed back to the sign-in page to try again

* We now wait before proceeding to guarded routes until authentication with Builder is complete (and reject if it fails)

It also does a bunch of cleanup, including removing some unused Redux code related to cookies (ensuring cookies are all handled in one place), simplifying the handling of the GitHub auth code, removing unused or no-longer-needed code, renaming a bit for clarity, and generally consolidating sign-in logic in fewer and more suitable places.

Signed-off-by: Christian Nunciato <cnunciato@chef.io>

Fixes #3882.

tenor-101592685

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/builder-web/app/app.component.html [hunk 1]
```diff
@@ -1,19 +1,14 @@
...
-    [removeNotification]="removeNotification"></hab-notifications>
-  <hab-header
-    [appName]="state.app.name"
-    [isUserNavOpen]="user.isUserNavOpen"
-    [isSignedIn]="user.isSignedIn"
-    [username]="user.username"
-    [avatarUrl]="user.gitHub.get('avatar_url')"
-    [signOut]="signOut"
-    [toggleUserNavMenu]="toggleUserNavMenu"></hab-header>
+  <hab-notifications [notifications]="state.notifications.all" [removeNotification]="removeNotification"></hab-notifications>
```
Comment (cnunciato) [inline components/builder-web/app/app.component.html]: I actually don't love this formatting, but my TSLint autocorrect settings did it automatically. I'm down to change this (although it's probably also a hint that components probably shouldn't take fifty inputs, anyway).

**Comment by cnunciato** [comment]
Just noticed a little oversight (username isn't showing in the dropdown). Fixing that up now.

**Comment by cnunciato** [comment]
All right, should be good to go now.

**Comment by cnunciato** [comment]
Now with more-nicer loading screen:

File: components/builder-web/app/sign-in-page/sign-in-page.component.ts [hunk 2]
```diff
@@ -37,9 +41,7 @@ export class SignInPageComponent implements OnInit, OnDestroy {
   }
 
   ngOnInit() {
```
Comment (elliott-davis) [inline components/builder-web/app/sign-in-page/sign-in-page.component.ts]: If this is empty can we just remove the function?

File: components/builder-web/app/sign-in-page/sign-in-page.component.ts [hunk 3]
```diff
@@ -14,15 +14,19 @@
...
 
 @Component({
   template: require('./sign-in-page.component.html')
 })
 export class SignInPageComponent implements OnInit, OnDestroy {
-  constructor(private store: AppStore) { }
+
+  constructor(private store: AppStore) {
+    store.dispatch(signOut(false));
+    this.store.dispatch(setLayout('centered'));
```
Comment (elliott-davis) [inline components/builder-web/app/sign-in-page/sign-in-page.component.ts]: I know you didn't add this but do we ever change the layout style?

File: components/builder-web/app/sign-in-page/sign-in-page.component.ts (hunk 2 above)
Comment (cnunciato) [reply components/builder-web/app/sign-in-page/sign-in-page.component.ts]: Definitely! I'll do that.

**Comment by cnunciato** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/296464094).

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/296464094).

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/296464094).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/296464094).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**